airbyte-integrations/base-java/src/main/java/io/airbyte/integrations/javabase/IntegrationCliParser.java

```diff
@@ -0,0 +1,191 @@
+package io.airbyte.integrations.javabase;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.OptionGroup;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Parses command line args to a type safe config object for each command type.
+ */
+public class IntegrationCliParser {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IntegrationCliParser.class);
+
+  private static final OptionGroup commandGroup = new OptionGroup();
+
+  static {
+    commandGroup.setRequired(true);
+    commandGroup.addOption(Option.builder()
+        .longOpt(Command.SPEC.toString().toLowerCase())
+        .desc("outputs the json configuration specification")
+        .build());
+    commandGroup.addOption(Option.builder()
+        .longOpt(Command.CHECK.toString().toLowerCase())
+        .desc("checks the config can be used to connect")
+        .build());
+    commandGroup.addOption(Option.builder()
+        .longOpt(Command.DISCOVER.toString().toLowerCase())
+        .desc("outputs a catalog describing the source's schema")
+        .build());
+    commandGroup.addOption(Option.builder()
+        .longOpt(Command.READ.toString().toLowerCase())
+        .desc("reads the source and outputs messages to STDOUT")
+        .build());
+    commandGroup.addOption(Option.builder()
+        .longOpt(Command.WRITE.toString().toLowerCase())
+        .desc("writes messages from STDIN to the integration")
+        .build());
+  }
+
+  public IntegrationConfig parse(final String[] args) throws IOException {
+    final Command command = parseCommand(args);
+    return parseOptions(args, command);
+  }
+
+  private static Command parseCommand(String[] args) throws IOException {
+    final CommandLineParser parser = new RelaxedParser();
+    final HelpFormatter helpFormatter = new HelpFormatter();
+
+    final Options options = new Options();
+    options.addOptionGroup(commandGroup);
+
+    try {
+      final CommandLine parsed = parser.parse(options, args);
+      return Command.valueOf(parsed.getOptions()[0].getLongOpt().toUpperCase());
+      // if discover, then validate, etc...
+    } catch (ParseException e) {
+      helpFormatter.printHelp("java-base", options);
+      throw new IOException(e);
+    }
+  }
+
+  private static IntegrationConfig parseOptions(String[] args, Command command) {
+
+    final Options options = new Options();
+    options.addOptionGroup(commandGroup); // so that the parser does not throw an exception when encounter command args.
+
+    switch (command) {
```

Comment:
https://argparse4j.github.io/ does a good job at defining sub-parser

Reply:
thanks! going to make a todo for this.